### ISSUE TITLE
squashfs becomes preferred image distribution method

### ISSUE THREAD

**Issue by shanegoff** [hpc/charliecloud#348]
# The Tar Method
Currently the Charliecloud preferred method of working with containers is to create an image tarball, then untar the image directly into tmpfs for all compute nodes.
This is a simple and well understood method, as most users are comfortable working with tarballs. This method does present some challenges. Distribution of container images at scale can be quite lengthy, and you must store your entire container image in memory.

# The Squash Filesystem
The squash file system(squashfs) provides a network mounted container image for use with Charliecloud. The primary advantages of this are reducing container image distribution time, and keeping more memory free on compute nodes.

The squashfs appears to be a clear winner for distribution of container images for Charliecloud. There are several different approaches for using the squashfs.
  * Kernel mounted squashfs
  * Fuse mounted squashfs

In testing the kernel mounted squashfs made the container available to nodes and had the fastest execution time for the squashfs methods. The execution time for kernel mounted squashfs is typically within seconds compared to tmpfs. As the squashfs does not need to be copied onto each node, the total run time was always lower than that of the current preferred method. Exectution time was relatively unhindered when using the kernel mounted squashfs. So the overhead experienced in the situation is less than that of untarring to nodes.

Fuse mounted squashfs through squashfuse has also been tested. Initially, only the high level API was tested and proved to be competitive with tmpfs for 'typical' HPC container workflow. The combined time of untar plus execution was greater than the execution time for the fuse mounted squashfs. It is note worthy that squashfuse suffers on very intense meta data operations such as filesystem walks, and reading the entire contents of the container. The kernel mounted squashfs saw much less impact in these extreme scenarios. Additionally, I learned of a low level fuse API available through squashfuse if you have the correct tools installed during the build process for squashfuse. The squashfuse_ll mount sees a smaller performance hit than squashfuse does. It performed very similarly in speed to the kernel mounted squashfs for execution, and handles metadata operations much faster than squashfuse. This could be an ideal middle ground, as we could have almost kernel mount performance without the kernel mount.

# What now

There are advantages, and disadvantages to the different mount methods. But I think moving to a squashfs solution as the new preferred method in some capacity for Charliecloud is the way to go.

## Support kernel mounted squashfs
The top performer is the kernel mounted squashfs. Adopting this method would also require the most investment, and carries the most risk. There would be a need for Charliecloud to be able to mount the squashfs as root in some way. Some mechanism to mount the squashfs would likely need to be built into Charliecloud. The kernel mounted squashfs has little need for optimizations, it works well with little or no effort on the users behalf.  This would also make it easier to support other filesystems in the future if need be. Squashfs is king today, but might not be in five years.

## Don't support kernel mounted squashfs
The low level fuse API through squashfuse performs very well without escalation to root. The overhead of using squashfuse_ll is lower than the overhead of untarring a whole image into memory as well. squashfuse_ll scales very well in testing up to 1024 nodes. Existing tools would make integrating squashfs into the existing Charliecloud workflow rather easy, but it could mean adding those tools as a dependency for Charliecloud. With default settings for the block size of squashfs, and Lustre default 1MiB stripe size squashfuse_ll performs almost as well as the 'optimized' performance, which is very narrowly faster. At scales larger than 1024 those tunings may become more important. The high level fuse API saw more notable improvement in the optimized form. It is worth noting I also did not try a squashfs block size smaller than the default of 128k. That could provide more speed improvements. One thing that does provide clear gains is the Lustre stripe pattern as you scale up.

Overall it seems as though squashfs is an improvement for container use with Charliecloud. Both squashfuse and kernel mounted squashfs reduce the total job time for execution of containers. As scale increases the time it takes to untar increases, but squashfuse_ll and kernel mounted squashfs do not show any note worthy degradation of execution time.

The following graphs show the total execution times for the Pynamic Benchmark, with the squashfs stored in Lustre using different configurations of squashfs block size, lustre stripe pattern, and Lustre stripe size.

Note I've taken the graphs down temporarily

# Questions
I would like input on this proposal, and am happy to answer any questions as best I can.

**Comment by olifre** [comment]
Just my 2 cents: We are using CVMFS ( https://cvmfs.readthedocs.io/en/stable/ ) as distribution method for containers. 
CVMFS is a read-only network file system, takes care of deduplication, can keep caches on all clients (or on a shared FS), and leverages standard HTTP infrastructure for data distribution. Especially if rebuilding containers frequently (e.g. for daily updates) or when keeping several container flavours, the deduplication really helps significantly. If we would use squashfs or tarballs, we could not rebuild all our containers each day (or would waste a lot of network bandwidth purely on that). With CMVFS, clients only need to fetch the parts they really use, and if they have fetched a full container once, they only need to fetch the things which actually do change and are used when a new version is committed. 

So in our workflow, CVMFS is the clear winner. We would not be happy with squashfs or distributing tarballs - we rather distribute the contents of the tarballs via CVMFS. 
Of course, I can imagine setups where squashfs will serve you better, but for us, CVMFS is the way to go. So I am not sure there is a single preferred method :wink: .

**Comment by Brofessional** [comment]
@olifre 
I agree there is no one ring to rule them all for container image distribution. I have never worked with CernVM-FS, and will look at the documentation to see what I'm missing out on. If you have some examples or details you could share with me regarding your container work flow it would certainly help me understand those tools better.

The squashfs is also a read-only network mounted file system. It has file deduplication, and has caching on clients as well. It is a compressed filesystem in a file, in this use case the file sits in Lustre to take advantage of parallelism and whatever caching Lustre may provide as well. Similar to the CVMFS, only the files required for running the container are read and stored in memory on compute nodes. However if a change needed to be made a user would need to create a new squashfs and prepare it for use on their system. In this scenario users don't have access to container registries from the web on the machine they're running on, only flattened container images they have prepared and made available for execution. The changes proposed would have no impact on the way Charliecloud currently works for running containers, it would only add support for the use of squashfs in some way. Charliecloud has some wrappers and tools to help container images distribution using tarballs, but supporting squashfs would create more options for users with very little potential support burden on the machines they plan on running on.

CVMFS may prove to have some benefits to users I do not understand. But Charliecloud could make containers faster for existing users that already use the workflow of preparing flattened images then using Charliecloud to make the image available and execute their application :)

**Comment by j-ogas** [comment]
I think presenting network mounted filesystems such as Squashfs/CernVM-FS as effective — perhaps superior — scaling alternatives to tarball distribution in our documentation is a good starting point. A developers guide preferably. 

That said, from my perspective, I am reluctant to embrace Squashfs as our *preferred* image distribution method namely because it:

1. Introduces a non-trivial dependency. At quick glance, Squashfs doesn't appear to be packaged in Fedora/EPEL. This implies that we will have to either distribute it (e.g., like `bats`, which is not preferred), or point users to third party webpages to download it, which further complicates our plans to package charliecloud for EPEL; 
2. Introductes a support burden (I'm looking at you, Vagrant); and 
3. I am unconvinced that the image filesystem size or unpacking time are big enough problems to justify introducing this solution as our standard. Although unpacking large images adds time, it is usually negligible compared to the runtime of the application of itself (assuming we are talking about a reasonably sized job). 

I see the benefits of squashfs in filesystem image deployment being most effective as a developers guide; an additional tool that users can leverage as opposed to a mandatory dependency we introduce.

**Comment by olifre** [comment]
> CVMFS may prove to have some benefits to users I do not understand.

The main advantage is that it can deduplicate across different containers, i.e. different container flavours, containers you rebuild daily (for updates) while keeping older versions around (for jobs still running) etc. Also, it does not need another filesystem (such as Lustre) for distribution - which may be an advantage or a disadvantage. 
Of course, it even needs it's own server infrastructure for that (which can in theory be a single server, but ideally would be more). We also use it for software kept outside of containers (and then being used via lmod), also here deduplication helps significantly. 

I tend to fully agree with @j-ogas . In many cases, squashfs or CVMFS (or maybe others I don't know of) will help significantly with container distribution, especially when your system scales to a significant size and containers are changed often. A mention in the documentation would for sure be helpful, but I am unsure using them by default is really a good idea due to the reasons outlined. 
That said, I believe that `squashfs-tools` are in EPEL:
https://fedoraproject.org/wiki/EPEL/epel7
and also `squashfuse` appears to be there:
https://centos.pkgs.org/7/epel-x86_64/squashfuse-0.1.102-1.el7.x86_64.rpm.html

**Comment by j-ogas** [comment]
@olifre Aha, thank you. Updating my original comment with the strikethrough of shame (ha).

**Comment by reidpr** [comment]
`ch-run` should work for any directory that looks like a Linux filesystem tree, regardless of how that directory got there. Thus, there are always going to be multiple workflows; this proposal is to promote squashfs as the "recommended" workflow instead of tarballs as currently.

I think this would make Charliecloud work *better* with CVMFS, because the recommended workflow would be a read-only mount instead of a read-write mount as we currently have with tarballs.

The major obstacle here is `ch-fromhost`, which currently assumes a writeable image, but we do have a workaround already, and there is a proposal to move `ch-fromhost` functionality into `ch-run` and make it work on read-only images (issue #286).

**Comment by mej** [comment]
Using `tar` files and directory trees keeps us most closely aligned with what OCI are doing, for better or [for worse](https://www.cyphar.com/blog/post/20190121-ociv2-images-i-tar).  There's definite value there.

I think keeping that the way it is is likely prudent, but we should certainly document a few "right ways" to get that directory structure distributed at scale., `squashfs` being perhaps the "premier" method?

**Comment by reidpr** [comment]
After the above and discussion in today's meeting, I think it's reasonable to start down this path. @Brofessional will articulate a squashfs-based workflow and the corresponding scripts in a separate PR.

I expect there will be additional issues and PRs working on different aspects of this.

**Comment by reidpr** [comment]
Depends on #414 and #441.

**Comment by reidpr** [comment]
This is more or less done.
